keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -0,0 +1,266 @@
+* published staging components can be successfully consumed by GOPATH-based consumers
+
+### Upgrade / Downgrade Strategy
+
+Not applicable
+
+### Version Skew Strategy
+
+Not applicable
+
+## Implementation History
+
+- 2019-03-19: Created
+
+## Alternatives
+
+* Continue using `godep` for vendor management. This is not viable for several reasons:
+  * The tool is unmaintained (the project readme states "Please use dep or another tool instead."), and we have had to make our own fork to work around some edge cases.
+  * There are significant performance problems (the pull-kubernetes-godeps CI job takes ~30 minutes to verify vendoring)
+  * There are significant functional problems (the tool cannot be run in some environments, gets confused by diamond dependencies, etc)
+
+* Use an alternate dependency mechanism (e.g. `dep`, `glide`). This is not preferred for several reasons:
+  * Some of the other dependency tools (like `glide`) are also unmaintained
+  * go modules are supported by the `go` tool, and have stronger support statements than independent vendoring tools
+  * Defining `go.mod` files for published kubernetes components is desired for interoperability with the go ecosystem
+
+* Move away from vendoring as part of the initial move to go modules
+  * To ensure reproducible builds based solely on the published repositories, vendoring is still necessary
+  * Moving away from vendoring is orthogonal to moving to go modules and could be investigated/pursued in the future if warranted.
+  * In go1.12.x, vendor-based builds are still the default when building a component located in the GOPATH, so producing components that work when built with go modules or with GOPATH+vendor maximizes interoperability
+
+* For versioning, switch to tagging major versions on every `kubernetes/kubernetes` release (similar to what client-go does), and use semantic import versioning.
+This remains a possibility in the future, but requires more tooling and consumer changes to accommodate rewritten imports,
+and doesn't fully allow multiple versions of kubernetes components to coexist as long as there are transitive non-module-based dependencies that change incompatibly over time.
+
+    * consumers
+      * GOPATH consumers
+        * import `k8s.io/apimachinery/...` (as they do today)
+        * are limited to a single copy of kubernetes libraries among all dependencies (as they are today)
+        * `go get` behavior (e.g. `go get client-go`):
+          * uses latest commits of transitive `k8s.io/...` dependencies (likely to break, as today)
+          * uses latest commits of transitive non-module-based dependencies (likely to break, as today)
+      * module-based consumers
+        * reference published modules versions as a consistent `vX.y.z` version (e.g. `v15.0.0`)
+        * import `k8s.io/apimachinery/v15/...` (have to rewrite kubernetes component imports on every major version bump)
```

Comment:
Conversely, if we actually did the right thing (as defined by go) and kept the old versions functioning correctly (e.g. v14.x.y itself imports v15 and adapts), then this would only be necessary when people actually wanted the new interface.

...I'm not claiming the demand go makes is reasonable, but if old packages stop working or stop working with new packages, we are not doing it the way they intended.

Reply:
I'm not necessarily opposed to adopting semantic import versioning, but I see that as a discrete step, and one that should wait at least until we don't have to worry about GOPATH consumers as well

Reply:
Agree.